internal/service/handlers/verify_passport.go

```diff
@@ -0,0 +1,175 @@
+package handlers
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math/big"
+	"net/http"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	zk "github.com/rarimo/zkverifier-kit"
+	"github.com/rarimo/zkverifier-kit/identity"
+
+	"github.com/rarimo/geo-auth-svc/internal/data"
+	"github.com/rarimo/geo-auth-svc/internal/jwt"
+	"github.com/rarimo/geo-auth-svc/internal/service/requests"
+	"gitlab.com/distributed_lab/ape"
+	"gitlab.com/distributed_lab/ape/problems"
+)
+
+func VerifyPassport(w http.ResponseWriter, r *http.Request) {
+	req, err := requests.NewVerifyPassport(r)
+	if err != nil {
+		Log(r).WithError(err).Debug("Bad request")
+		ape.RenderErr(w, problems.BadRequest(err)...)
+		return
+	}
+
+	log := Log(r).WithFields(map[string]any{
+		"user.nullifier":    req.Data.ID,
+		"user.anonymous_id": req.Data.Attributes.AnonymousId,
+	})
+
+	nullifier := req.Data.ID
+	anonymousID := req.Data.Attributes.AnonymousId
+	proof := req.Data.Attributes.Proof
+
+	claim := Claim(r)
+	if claim == nil || claim.Type != jwt.AccessTokenType || claim.Nullifier != nullifier {
+		ape.RenderErr(w, problems.Unauthorized())
+		return
+	}
+
+	gotSig := r.Header.Get("Signature")
+	wantSig, err := calculatePassportVerificationSignature(SigVerifier(r), req.Data.ID, anonymousID)
+	if err != nil { // must never happen due to preceding validation
+		Log(r).WithError(err).Error("Failed to calculate HMAC signature")
+		ape.RenderErr(w, problems.InternalError())
+		return
+	}
+
+	if gotSig != wantSig {
+		log.Warnf("Passport verification unauthorized access: HMAC signature mismatch: got %s, want %s", gotSig, wantSig)
+		ape.RenderErr(w, problems.Forbidden())
+		return
+	}
+
+	byNullifier, err := UsersQ(r).FilterByNullifier(nullifier).Get()
+	if err != nil {
+		log.WithError(err).Error("Failed to get user by nullifier")
+		ape.RenderErr(w, problems.InternalError())
+		return
+	}
+
+	if byNullifier != nil && byNullifier.IsProven {
+		log.Warnf("User %s already proven", nullifier)
+		ape.RenderErr(w, problems.TooManyRequests())
+		return
+	}
+
+	byAnonymousID, err := UsersQ(r).FilterByAnonymousID(anonymousID).Get()
+	if err != nil {
+		log.WithError(err).Error("Failed to get user by anonymous ID")
+		ape.RenderErr(w, problems.InternalError())
+		return
+	}
+
+	if (byNullifier == nil || byAnonymousID == nil || byNullifier.Nullifier != byAnonymousID.Nullifier) && (byNullifier != nil || byAnonymousID != nil) {
```

Comment:
maybe, we just drop this and have 4 separate conditions in switch/case? Also everything is 409, try putting this intt function and render once